addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,304 @@
+        }
+        if (!definition.get('lifecycleHooks').has(name)) {
+            throw new Error(`Cannot patch lifecycle hook "${name}" on model "${modelName}": hook does not exist yet.`);
+        }
+        const hookBeforePatch = definition.get('lifecycleHooks').get(name);
+        definition.get('lifecycleHooks').set(name, function (...args) {
+            this._super = hookBeforePatch;
+            return handler.call(this, ...args);
+        });
+    }
 }
 
-//------------------------------------------------------------------------------
-// Export
-//------------------------------------------------------------------------------
+/**
+ * Overrides the model methods of the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Model methods to be overriden.
+ */
+export function patchModelMethods(modelName, modelMethods) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot patch model methods on model "${modelName}": model must be registered before being patched.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [name, method] of Object.entries(modelMethods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch model method "${name}" on model "${modelName}": "${name}" must be a function.`);
+        }
+        if (!definition.get('modelMethods').has(name)) {
+            throw new Error(`Cannot patch model method "${name}" on model "${modelName}": method does not exist yet.`);
+        }
+        const methodBeforePatch = definition.get('modelMethods').get(name);
+        definition.get('modelMethods').set(name, function (...args) {
+            this._super = methodBeforePatch;
+            return method.call(this, ...args);
+        });
+    }
+}
 
-export {
-    registerClassPatchModel,
-    registerFieldPatchModel,
-    registerInstancePatchModel,
-    registerIdentifyingFieldsPatch,
-    registerNewModel,
-    registry,
-};
+/**
+ * Overrides the record methods of the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Record methods to be overriden.
+ */
+export function patchRecordMethods(modelName, recordMethods) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot patch record methods on model "${modelName}": model must be registered before being patched.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [name, method] of Object.entries(recordMethods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch record method "${name}" on model "${modelName}": "${name}" must be a function.`);
+        }
+        if (!definition.get('recordMethods').has(name)) {
+            throw new Error(`Cannot patch record method "${name}" on model "${modelName}": method does not exist yet.`);
+        }
```

Comment:
I feel like there is a combinatorial explosion with all of those checks.

There should be one method to check that "a name is free" in model and one in record, the rest should just be calling it.
If you want a nice error message, you can catch the exception up and add the missing context into its message.

Reply:
And same reasoning for all other types of check, "model must be registered" should only appear once in the file